joomla/joomla-cms
# [5.2] SEF: Enforcing correct SEF URL

### Summary of Changes
Joomla has been improving its SEO performance constantly and one issue which is still open is, that you can reach the same page via different URLs. This means that you can reach the same article both via the wanted URL, but also via the `/component/content/article/<id>` URL or even via `index.php?option=com_content&view=article&catid=<catid>&id=<id>`. 

This PR adds a new feature which redirects URLs with query parameters to their SEF variant. The router parses the URL and if the original URL contained query parameters, it tries to create a new URL from the information it parsed and if that new URL does not contain any query parameters anymore, it does a redirect to that URL. This will remove a lot of the double URLs for one content item. This only happens for GET requests.

It should be noted, that this does NOT remove every case of duplicate content. For example additional URL parameters are still kept.

I'd like to thank [ithelps Digital](https://www.ithelps-digital.com/) for sponsoring this feature.


### Testing Instructions
1. Apply the patch
2. Use some content to test this with, for example the testing sample data.
3. Go to `/park-blog/first-blog-post` in your site when you use the testing sample data
4. Add a `?id=18` to the URL and see that you actually get the "Second Blog Post" displayed instead of the first one.
5. Enable the option in the SEF plugin.
6. Reload the frontend page and see that you are redirected to `/park-blog/second-blog-post`.



### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: https://docs.joomla.org/Search_Engine_Friendly_URLs
